### PR TITLE
Add support for rubix data decoding

### PR DESCRIPTION
**Pull request by lmtan91**
- Rewrite C++ decoder in rubix-data-encoding project
- Add unittest rubix_test.go for decoder

### CODE REVIEW

File: decoder/rubix.go [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+
+func DecodeRubix(data string, _ *LoRaDeviceDescription) (*CommonValues, interface{}) {
+	bytes, err := hex.DecodeString(data)
+	if err != nil {
+		log.Errorf("error decoding rubix hex data")
+		return nil, nil
+	}
+	sd := NewSerialDataWithBuffer(bytes)
+
+	// TODO: implement decode loop
```
Comment (Shiny380) [inline decoder/rubix.go:402]: @lmtan91 See JS example of decode loop implementation here: https://github.com/NubeIO/rubix-data-encoding/blob/master/js/decode-ES5.js#L194-L219

File: decoder/rubix.go [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+	MDK_DOUBLE:           {DATAPOINT, 0, 0, 0, 8},
+}
+
+func DecodeRubix(data string, _ *LoRaDeviceDescription) (*CommonValues, interface{}) {
+	bytes, err := hex.DecodeString(data)
+	if err != nil {
+		log.Errorf("error decoding rubix hex data")
+		return nil, nil
+	}
+	sd := NewSerialDataWithBuffer(bytes)
```
Comment (Shiny380) [inline decoder/rubix.go:400]: before doing this, trim the data like so: https://github.com/NubeIO/rubix-io-fw/blob/master/src/core/comms/loraRaw/loraRawProtocolHandler.cpp#L72-L73

File: decoder/rubix.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+
+	return nil, nil
+}
+
+func GetPointsStructRubix() interface{} {
+	return Rubix{}
+}
+
+func CheckPayloadLengthRubix(data string) bool {
+	return true
```
Comment (Shiny380) [inline decoder/rubix.go:412]: see length checks here: https://github.com/NubeIO/rubix-io-fw/blob/master/src/core/comms/loraRaw/loraRawProtocolHandler.cpp#L87-L90

**Review by Shiny380 — CHANGES_REQUESTED** [review]

File: decoder/rubix.go (hunk 2 above)
Comment (lmtan91) [reply decoder/rubix.go:400]: I am not able to access to this: rubix-io-fw

**Comment by Shiny380** [comment]
@lmtan91 you should be able to mock up test code based on [this code](https://github.com/NubeIO/module-core-loraraw/blob/master/pkg/app.go#L108C2-L115C3) using the following hex data for testing:
```
8AC0311700A9210044004400440044004400440044004400480148014801480148014801480148014427
```
- `8AC03117` is the device address
- `00A9` is the nonce and data length
- the rest is data
- the last 4 bytes are the SNR and RSSI

File: decoder/rubix.go (hunk 3 above)
Comment (Shiny380) [reply decoder/rubix.go:412]: I just checked and the driver that outputs the hex data already strips off the MAC so you can exclude that

Comment (Shiny380) [reply decoder/rubix.go:412]: hex output driver here: https://github.com/NubeIO/driver-lora/blob/master/src/loRaRaw.py#L57-L75

Comment (Shiny380) [reply decoder/rubix.go:532]: @lmtan91 this one is not resolved, Check that the length byte in the data is correct for the payload.
i.e. `AABBCCDD000511223344558899` `05` is the length byte and `1122334455` = 5 bytes so this is correct (`8899` is the SNR and RSSI)

File: decoder/rubix.go [hunk 4]
```diff
@@ -331,95 +373,169 @@ func DecodeRubix(data string, devDesc *LoRaDeviceDescription, device *model.Devi
...
 
 	serialData := NewSerialDataWithBuffer(payload)
 
-	var float_count = 0
-
 	for canDecode(serialData) {
-		pos := uint8(0)
-		header := getHeader(serialData, &pos)
+		header := getHeader(serialData, &position)
+		hasPos := hasPositionalData(serialData)
```
Comment (Shiny380) [inline decoder/rubix.go:444]: this can be moved outside of loop as its defined for the entire payload

File: decoder/rubix.go [hunk 5]
```diff
@@ -9,10 +9,13 @@ import (
 	"unsafe"
 
 	"github.com/NubeIO/nubeio-rubix-lib-models-go/model"
+	log "github.com/sirupsen/logrus"
 )
 
 const HEADER_BIT_COUNT = 6
 
+var position uint8 = 0
```
Comment (Shiny380) [inline decoder/rubix.go:17]: move to local variable of decode func instead of global. This is also never reset to 0 for subsequent payloads coming in

**Review by Shiny380 — CHANGES_REQUESTED** [review]
@Tan see two change requests and also:
1. Can you please add more than one test that get passed to the decoder in the same instantiation, without resetting the decoder
2. Can you please also make sure you have one positional packet to test and one non-positional packet

**Review by Shiny380 — APPROVED** [review]

**Pull request merged**